todo tool feature added to tools

gayatrik-dev opened the pull request:
This PR addresses #[1966](https://github.com/arc53/DocsGPT/issues/1966).

### Features Implemented

The tool provides complete CRUD support via the following actions:
- todo_create: Add a new todo item with optional description, due date, and metadata.
- todo_get: Retrieve a specific todo by its unique todo_id.
- todo_list: List all todos created by a specific user-tool pair.
- todo_update: Update title, description, status, due date, or metadata of an existing todo.
- todo_delete: Remove a specific todo by todo_id.

### Technical Highlights
- Todos are uniquely identified using UUID-based todo_ids.
- MongoDB-backed with appropriate indexes on todo_id, user_id, and tool_id to ensure efficient queries.
- Tool configuration supports an optional tool_id for contextual separation; defaults to default_{user_id} if none provided.

### Tests
- Added a comprehensive suite of unit tests using a fake in-memory MongoDB collection:
- Covers creation, retrieval, listing, updating, deletion, and isolation across tool_ids and users.

dartpain commented:
Todo_id should correspond to a  tool _id in the user_tools collection.

todo_id = str(uuid.uuid4())

Please check out https://github.com/arc53/DocsGPT/blob/main/application/agents/tools/memory.py
and https://github.com/arc53/DocsGPT/blob/main/application/agents/tools/notes.py

They handle it well there.

On tools init you should get tool_config param:
and use it to store tool_id

```
if tool_config and "tool_id" in tool_config:
            self.tool_id = tool_config["tool_id"]
```

gayatrik-dev commented:
Hi, I hope this addresses your concerns — the refactor supports either many todos within a single tool or multiple tools each with their own todos for a given user.

gayatrik-dev commented:
How about we introduce a counters collection in MongoDB to track the current sequence number?
Then for each new todo, we use find_one_and_update with $inc to atomically generate the next todo_id.

gayatrik-dev commented:
Hi, this isn't merged yet. any blockages?
